pipeline_branching.md

```diff
@@ -0,0 +1,159 @@
+# Pipeline Branching
+
+## Goal
+
+Allow users to have multiple, named versions of a pipeline (branches)
+running concurrently.
+
+## Use Case:
+
+The primary use case for this is to give users a good way to have `dev`,
+`staging` and `prod` versions of pipelines. There are a ton of use cases for
+this feature, but so far I haven't encountered one that presented challenges
+that the `dev`, `staging` `prod` use case doesn't, so for this document we'll
+just stick to that for inspiration.
+
+## PPS API changes:
+The pipeline message will be extended to include a `Branch` field, which
+is a string:
+
+```proto
+message Pipeline {
+  string name = 1;
+  string branch = 2;
+}
+```
+
+In pachctl this is implemented with the familiar syntax of `pipeline@branch`.
+
+An empty `branch` field is equivalent of a value of `"master"`.
+
+Different branches of the same pipeline are effectively different pipelines. That means:
+- They output to different branches, with names that match their branch names.
+- You can delete, update, pause, etc. one without pausing the other.
+- They can take different inputs, have different resource requests, etc.
+- They run in different pods, and their success / failure is independent.
+
+There are a few ways in which they're not quite totally distinct:
+- They consolidate in list pipeline, this one's mostly cosmetic, I'm
+imagining something that looks like this:
+```
+NAME     BRANCHES     INPUT    CREATED    STATE / LAST JOB  DESCRIPTION
+pipeline dev, staging input:/* 6 days ago running / success
+```
+- They share datums for deduplication, if you want a dev version to
+reprocess, you need to create it with `--reprocess`.
+- They can't take each other as input. Implementation wise there's no
+particular reason for this, but it seems very weird and I can't think of any
+reason you'd want to do this that there isn't a better way of accomplishing.
+
+### A few examples of what syntax would look like for this:
+Creating a dev branch:
+
+```
+extract pipeline pipeline >pipeline.json
+# modify pipeline.json, setting Pipeline.Branch to "dev"
+create pipeline -f pipeline.json
+```
+
+List jobs from dev branch:
+
+```
+list job -p pipeline@dev
+```
+
+Get files from dev branch (all other file commands work similarly):
+
+```
+get file pipeline@dev:/file
+```
+
+Promote dev branch to master
+
+```
+# modify pipeline.json, setting Pipeline.Branch to ""
+create pipeline -f pipeline.json
```

Comment:
Since master already exists, will this be create or update pipeline?

Reply:
Either works actually. `update` supports upsert semantics, so you can use `update` wherever you can use `create`. The distinguishing feature of `create` is that it won't overwrite something that's already there, because this is a different branch from `master` there's nothing it will overwrite, so you can use `create` as well, and generally you should, since it removes the risk of accidentally overwriting something you want.